src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,235 @@
+  /** Holds the tuples that are ready for release. */
+  private transient TupleBatchBuffer buffer;
+  /** Which line of the file the scanner is currently on. */
+  private long lineNumber = 0;
+
+  private final boolean isLastWorker;
+  private final long fileSize;
+
+  private long byteOverlap = 10;
+  private long startByteRange;
+  private long endByteRange;
+  long workerID;
+
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * The logger for debug, trace, etc. messages in this class.
+   */
+  private static final org.slf4j.Logger LOGGER = org.slf4j.LoggerFactory.getLogger(CSVFileScanFragment.class);
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers) {
+    this(filename, schema, workerID, totalWorkers, null, null, null, null);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers) {
+    this(source, schema, workerID, totalWorkers, null, null, null, null);
+  }
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers,
+      final Character delimiter) {
+    this(new FileSource(filename), schema, workerID, totalWorkers, delimiter, null, null, null);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers,
+      final Character delimiter) {
+    this(source, schema, workerID, totalWorkers, delimiter, null, null, null);
+  }
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers,
+      @Nullable final Character delimiter, @Nullable final Character quote, @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this(new FileSource(filename), schema, workerID, totalWorkers, delimiter, quote, escape, numberOfSkippedLines);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers,
+      @Nullable final Character delimiter, @Nullable final Character quote, @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this.source = (AmazonS3Source) Preconditions.checkNotNull(source, "source");
+    this.schema = Preconditions.checkNotNull(schema, "schema");
+
+    this.delimiter = MoreObjects.firstNonNull(delimiter, CSVFormat.DEFAULT.getDelimiter());
+    this.quote = MoreObjects.firstNonNull(quote, CSVFormat.DEFAULT.getQuoteCharacter());
+    this.escape = escape != null ? escape : CSVFormat.DEFAULT.getEscapeCharacter();
+    this.numberOfSkippedLines = MoreObjects.firstNonNull(numberOfSkippedLines, 0);
+    this.workerID = workerID;
+
+    isLastWorker = workerID == totalWorkers;
+    fileSize = ((AmazonS3Source) source).getFileSize();
+
+    if (fileSize < 10 && workerID == 1) {
```

Comment:
What about cases where there are more workers than bytes in the file? Also, weren't we going to have a configurable minimum partition size (could just be a constant in `edu.washington.escience.myria.MyriaConstants` for now, but should definitely not be a magic number)? That, however, would determine the number of workers on which the file is partitioned, so would need to be done in the coordinator (which could also calculate S3 object size). The alternative would be to have each worker perform the partitioning calculation independently to determine whether it should store any data at all, in which case it might have a "dummy" partition in the catalog with no data, which would incur RPC overhead on every query for no benefit (assuming that there's no way to avoid creating an unnecessary partition without coordination). Is that price worth the benefit of avoiding coordination?